src/libraries/Microsoft.Extensions.Diagnostics/src/Metrics/ListenerSubscription.cs

```diff
@@ -165,36 +165,39 @@ internal static bool RuleMatches(InstrumentRule rule, Instrument instrument, str
 
             // Meter
 
-            var ruleMeterName = rule.MeterName.AsSpan();
-            // Don't allow "*" anywhere except at the end.
-            var starIndex = ruleMeterName.IndexOf('*');
-            if (starIndex != -1 && starIndex != ruleMeterName.Length - 1)
+            // The same logic as Microsoft.Extensions.Logging.LoggerRuleSelector.IsBetter for category names
+            var meterName = rule.MeterName;
+            if (meterName != null)
             {
-                return false;
-            }
-            // Rule "System.Net.*" matches meter "System.Net" and "System.Net.Http"
-            if (ruleMeterName.EndsWith(".*".AsSpan(), StringComparison.Ordinal))
-            {
-                ruleMeterName = ruleMeterName.Slice(0, ruleMeterName.Length - 2);
-            }
-            // System.Net* matches System.Net and System.Net.Http
-            else if (starIndex != -1)
-            {
-                ruleMeterName = ruleMeterName.Slice(0, ruleMeterName.Length - 1);
-            }
+                const char WildcardChar = '*';
 
-            // Rule "" matches everything
-            if (ruleMeterName.IsEmpty)
-            {
-                return true;
+                int wildcardIndex = meterName.IndexOf(WildcardChar);
+                if (wildcardIndex != -1 &&
+                    meterName.IndexOf(WildcardChar, wildcardIndex + 1) != -1)
+                {
+                    throw new InvalidOperationException(SR.MoreThanOneWildcard);
+                }
+
+                ReadOnlySpan<char> prefix, suffix;
+                if (wildcardIndex == -1)
+                {
+                    prefix = meterName.AsSpan();
+                    suffix = default;
+                }
+                else
+                {
+                    prefix = meterName.AsSpan(0, wildcardIndex);
+                    suffix = meterName.AsSpan(wildcardIndex + 1);
+                }
+
+                if (!instrument.Meter.Name.AsSpan().StartsWith(prefix, StringComparison.OrdinalIgnoreCase) ||
+                    !instrument.Meter.Name.AsSpan().EndsWith(suffix, StringComparison.OrdinalIgnoreCase))
```

Comment:
what the behavior will be if have like, meter name `ABCDE` and have searching for `ABC*CDE`. This search should fail but with current code it will succeed I guess. Maybe we need to check the length of prefix and suffix against the meter name?

Reply:
Since we're going for complete compatibility with logging's matching I'd rather not adjust the behavior for corner cases like this. We can revisit this in 9 for both of them if you want.

Reply:
Would it be a good idea to stick with the wrong behavior? This will be compatibility issue if we don't do it from now. I prefer doing it but I'll leave it to you if you think otherwise.

Reply:
I agree its weird, but my rationale has remained that if people aren't complaining about it via ILogger then we should remain compatible. If someone did complain about it I assume we'd want to change it uniformly, not change it just for metrics.